### PR TITLE
CLOUDP-111968: Watch for changes in CA certificate

### PR DESCRIPTION
**Pull request by mircea-cosbuc**
### All Submissions:

This patch triggers a reconciliation loop when the CA certificate changes. 
**NOTE**: The server certificates should be valid under both the old and new CA certificate chains for a zero-downtime update.
This also closes https://github.com/mongodb/mongodb-kubernetes-operator/issues/874.

* [ ] Have you opened an Issue before filing this PR?
* [ ] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

### CODE REVIEW

**Review by rodrigovalin — APPROVED** [review]
LGTM

File: controllers/mongodb_tls.go [hunk 1]
```diff
@@ -91,23 +73,35 @@ func (r *ReplicaSetReconciler) validateTLSConfig(mdb mdbv1.MongoDBCommunity) (bo
 	// Watch certificate-key secret to handle rotations
 	r.secretWatcher.Watch(mdb.TLSSecretNamespacedName(), mdb.NamespacedName())
 
+	// Watch CA certificate changes
+	if mdb.Spec.Security.TLS.CaCertificateSecret != nil {
+		r.secretWatcher.Watch(mdb.TLSCaCertificateSecretNamespacedName(), mdb.NamespacedName())
+	} else {
+		r.configMapWatcher.Watch(mdb.TLSConfigMapNamespacedName(), mdb.NamespacedName())
+	}
+
```
Comment (irajdeep) [inline controllers/mongodb_tls.go:82]: Q: Can both be specified, In this case we would like to prefer Secret over ConfigMap right?

File: controllers/mongodb_tls.go [hunk 2]
```diff
@@ -162,6 +156,51 @@ func getPemOrConcatenatedCrtAndKey(getter secret.Getter, mdb mdbv1.MongoDBCommun
...
 }
 
+func getCaCrt(cmGetter configmap.Getter, secretGetter secret.Getter, mdb mdbv1.MongoDBCommunity) (string, error) {
+	var caResourceName types.NamespacedName
+	var caResourceType string
+	var caData map[string]string
+	var err error
+	if mdb.Spec.Security.TLS.CaCertificateSecret != nil {
+		caResourceName = mdb.TLSCaCertificateSecretNamespacedName()
+		caResourceType = "Secret"
```
Comment (irajdeep) [inline controllers/mongodb_tls.go:166]: nit: `caResourceType` can be made an enum?

File: deploy/e2e/role.yaml [hunk 3]
```diff
@@ -173,6 +173,7 @@ rules:
   - watch
   - create
   - update
+  - patch
```
Comment (irajdeep) [inline deploy/e2e/role.yaml:176]: Why is the patch permission needed now?

**Review by irajdeep — COMMENTED** [review]
LGTM -- with some minor comments inline

File: controllers/mongodb_tls.go (hunk 1 above)
Comment (mircea-cosbuc) [reply controllers/mongodb_tls.go:82]: This was the current behaviour, yes, the secret takes precedence.

File: deploy/e2e/role.yaml (hunk 3 above)
Comment (mircea-cosbuc) [reply deploy/e2e/role.yaml:176]: The e2e test which is extending the CA certificate is performing a patch on a `Certificate` resource.

File: controllers/mongodb_tls.go (hunk 2 above)
Comment (irajdeep) [reply controllers/mongodb_tls.go:166]: I think in that case it's better to get rid of it.

**Review by irajdeep — APPROVED** [review]
LGTM -- nice work.

**Pull request merged**